File: bit_manipulation/0187_Repeated_DNA_Sequences/solution.cpp
#include <vector>
#include <string>
#include <cstdint>
#include <array>
#include <string.h>
#include <unordered_map>
#include <iostream>
#include <climits>

using namespace std;

class Solution {
public:
    vector<string> operator()(string s)  {
        return findRepeatedDnaSequences(s);
    }

    void makeHash(int &hash, int letter) {
       // A 1 C 3 G 7 T 20 after & 0x3f, after >> 1 (middle + left), after &0b11 (middle) :
       // 0b00 00  1
       // 0b00 01  1
       // 0b00 11  1
       // 0b10 10  0
        int c = ((letter & 0x3f) >> 1)  & 0b11;
        hash &= 0b00111111111111111111; // avoid overflow
        hash = (hash << 2 | c); // add current letter information to value of the hash
    }

    vector<string> findRepeatedDnaSequences(string s) {
        vector<string> res = {};
        if (s.size() <= 10)
            return res;
        int hashTable[1 << 20] = {0};

        int hash = 0;
        for (int i = 0; i < 10; i++)
            makeHash(hash, s[i]);
        hashTable[hash]++;
        for (int i = 10; i < s.size(); i++) {
            makeHash(hash, s[i]);
            if (hashTable[hash]++ == 1)
                res.push_back(s.substr(i - 9, 10));
        }
        return res;
    }

    vector<string> findRepeatedDnaSequences2(string s) { // the easy way
            vector<string> res = {};
            if (s.size() < 10)
                return res;
            const char *c = s.c_str();
            unordered_map<string, int> m;
            char buf[s.size() + 1];
            (void)buf;
            memcpy(buf, c, 10);
            m[string(&buf[0], 10)]++;

            for (int i = 10; i < s.size(); i++) {
                memcpy(&buf[i], &s[i], sizeof(char));
                buf[i + 1] = '\0';
                string current = string(&buf[i - 9], 10);
                if (m[current]++ == 1 )
                    res.push_back(&buf[i - 9]);
            }
            return res;
     }
};

File: bit_manipulation/0187_Repeated_DNA_Sequences/solution.rs
impl Solution {
    fn make_hash(hash: &mut i32, letter: i32) {
        let c:i32 = ((letter & 0x3f) >> 1) & 0b11;
        *hash &= 0b00111111111111111111;
        *hash = (*hash << 2 | c);
    }
    pub fn find_repeated_dna_sequences(s: String) -> Vec<String> {
         let mut res:Vec<String>  = Vec::new();

        if (s.chars().count() <= 10) {
            return res;
        }

        let mut hashTable:[i32;1 << 20] = [0; 1 << 20];
        let mut hash:i32 = 0;
        for i in 0..10 {
            Solution::make_hash(&mut hash, s.as_bytes()[i] as i32);
        }
        hashTable[hash as usize]+=1;

        for i in 10..s.chars().count() {
            Solution::make_hash(&mut hash, s.as_bytes()[i] as i32);
            if (hashTable[hash as usize] == 1) == true {
                res.push(s.chars().skip(i-9).take(10).collect());
            }
            hashTable[hash as usize] += 1;
        }

        return res;
    }
}